src/Graphics/Image/Processing/Noise.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE BangPatterns #-} 
+
+module Graphics.Image.Processing.Noise where
+
+import Control.Monad (forM_)
+import Control.Monad.ST
+import System.Random
+
+import Prelude as P hiding (subtract)
+import Graphics.Image.Interface as I
+import Graphics.Image
+import Graphics.Image.Types as IP
+import Graphics.Image.ColorSpace (X)
+
+-- | Helper function for generating a list of random co-ordinates.
+randomCoords :: StdGen -> Int -> Int -> [(Int,Int)]
+randomCoords a width height = (rnx1, rny1) : randomCoords g2 width height
+  where
+    (rnx1, g1) = randomR (1, width) a
+    (rny1, g2) = randomR (1, height) g1
+
+-- | Salt and pepper noise or impulse noise is a form of noise seen on images.
+-- It is mainly caused by sharp and sudden disturbances in the image signal. 
+-- 
+-- 'snp' generates this particular type of noise by introducing a sparse 
+-- distribution of white and black pixels in the input image. The level or intensity
+-- of the noise to be introduced is a parameter of this method and is scaled 
+-- between 0 and 1, that is the input Noise Intensity has a domain : (0, 1).
+-- 
+-- <<images/yield.jpg>>   <<images/yield_snp.png>>
+--
+-- Usage :
+-- 
+-- >>> img <- readImageY VU "images/yield.jpg"
+-- >>> input1 <- getLine
+-- >>> input2 <- getLine
+-- >>> input3 <- getLine
+-- >>> g <- newStdGen
+-- >>> let thetaSz = (P.read input1 :: Int)
+-- >>> let distSz = (P.read input2 :: Int) 
+-- >>> let noiseLevel = (P.read input3 :: Float)
+-- >>> let snpImage :: Image VU Y Word16
+-- >>>     snpImage = snp img thetaSz distSz noiseLevel g
+-- >>> writeImage "images/yield_snp.png" (toImageRGB snpImage)
+--
+snp
+  :: forall arr e cs . (MArray arr Y Double, IP.Array arr Y Double, IP.Array arr Y Word16, MArray arr Y Word16)
+  => Image arr Y Double
+  -> Int    -- ^ width of output image
```

Comment:
It is incorrect to ask for width and height as arguments for this function, we aren't changing image dimensions are we?